---
title: "Microbial enrichment and storage for metagenomics of vaginal,skin and saliva samples"
author: "Sarah Ahannach"
date: "August 30, 2021"
output: html_document
---

This Rmarkdown document describes the data analysis for the paper "Microbial enrichment and storage for metagenomics of vaginal,skin and saliva samples" by Sarah Ahannach, Lize Delanghe, Irina Spacova, Stijn Wittouck, Wannes Van Beeck, Ilke De Boeck, Sarah Lebeer. All data visualizations and summary tables shown in the paper are generated by this document. You can read this document to get a more detailed overview of the data analysis, or you can run the code yourself and play around with it if you want to explore the data on your own. The easiest way to run the code is to use [RStudio](https://www.rstudio.com/products/rstudio/#Desktop). 

--------
Content
--------
- Set-up
- Part 1: 16S rRNA gene amplicon sequencing analysis
- Part 2: Shallow metagenomic shotgun sequencing analysis

--------------------------------------------------------------------------------------------------------------------------------------------------
Set-up
--------------------------------------------------------------------------------------------------------------------------------------------------

You can install these packages by uncommenting and running the following commands:
```{r}
#install.packages("tidyverse")
#devtools::install_github("SWittouck/tidyamplicons")
#install.packages("rmarkdown")
#install.packages("devtools")
#install.packages("Hmisc")
#install.packages("ggpubr")
#install.packages("vegan")
#install.packages("ape")
#install.packages("ggtree")
#install.packages("phylogram")
#install.packages("ggrepel")
```

Settings for rendering of this markdown file:
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load the required R packages: 
```{r}
library(tidyverse)
library(tidyamplicons)
library(readr)
library(Hmisc)
library(ggpubr)
library(vegan)
library(ape)
library(ggtree)
library(phylogram)
library(ggrepel)
```

--------------------------------------------------------------------------------------------------------------------------------------------------
PART 1
--------------------------------------------------------------------------------------------------------------------------------------------------

## Data preparation

We start by loading the data. All data is bundled in a "tidyamplicons" object: it contains the read counts of all taxa (ASVs) in all samples, the metadata of the samples (sampling location, swab, method, timepoint,...) and metadata of the taxa (taxonomic classification and interpretable taxon names). 
```{r}
load("../data/isala_amplicon.rda")
```

Next we calculate the library concentration by making a new column: lib_size/pooled volume = concentration. But don't forget to first add lib_size
```{r}
run <- add_lib_size(run)
run <- mutate_samples (run, concentration = lib_size / pooled_volume)
#View(run$samples)
```

## ASV filtering

### Too long reads

General exploration
```{r}
get_numbers(run)
```

Plot length of ASVs:
```{r,fig.height= 3}
run$taxa <- run$taxa %>%
  mutate(length = str_length(taxon))
run$taxa %>%
  count(length) %>%
  rename(n_asvs = n) %>%
  ggplot(aes(x =length, y = n_asvs)) +
  geom_bar(stat = "identity") +
  scale_y_log10()

run %>%
  add_lib_size() %>%
  samples()%>%
  ggplot(aes(fct_reorder(sample,lib_size),lib_size)) +
  geom_col()

#if you want descending
run %>%
  add_lib_size() %>%
  samples()%>%
  ggplot(aes(fct_reorder(sample,lib_size, .desc = T),lib_size)) +
  geom_col()
```

Filter out ASVs that are too long:
```{r}
run$taxa <- run$taxa %>%
  filter_taxa(length < 300) %>%
  update_lib_sizes(step = "removing_too_long")

get_numbers(run)
```


### Non-bacterial reads

```{r, message=TRUE, fig.height=3}
run <- add_lib_size(run)
run <- mutate_samples(run, libmcb = lib_size) # libmcb is now the original amount of all reads, including non-bacterial

run$taxa <- run$taxa %>%
  filter_taxa(kingdom == "Bacteria") %>%
  filter_taxa(family !="Mitochondria" | is.na(family)) %>%
  filter_taxa(class != "Chloroplast" | is.na(class))
#add | is.na(...) --> because you don't want to filter our the NA's
run <- add_lib_size(run)

run <- mutate_samples(run, nonbacterial=(libmcb-lib_size)/libmcb)
```

```{r}
run <- update_lib_sizes(run, step = "removing_non_bacterial")
get_numbers(run)
```

Plot non-bacterial per sample
```{r,fig.height=5}
ggplot(run$samples, aes(x = sample_id, y=nonbacterial)) +
  geom_jitter(size = 2) +
  geom_boxplot(alpha = 0.1) +
  theme_bw()
```

Calculate non-bacterial in samples
```{r}
run_mean <- run$samples %>%
  filter(!is.na(nonbacterial))
mean(run_mean$nonbacterial)
```


### Quality control of library

```{r,fig.height=3}
ggplot(run$lib_sizes, aes(x = lib_size)) +
  scale_x_log10() +
  geom_density() +
  geom_rug()
```

```{r}
ggplot(run$lib_sizes, aes(x = lib_size,  y = sample_id)) +
    geom_jitter(alpha = 0.2)
```

```{r}
ggplot(run$samples, aes(x = lib_size,  y = sample_id)) +
  geom_jitter(alpha = 0.2)
```

```{r}
ggplot(run$samples, aes(x = sample_id, y = lib_size)) +
  geom_jitter(size = 2) +
  geom_boxplot (alpha = 0.1) +
  scale_y_log10() +
  theme_bw()
```

```{r}
run %>%
  add_lib_size() %>%
  samples() %>%
  ggplot(aes(x = sample_id, y = lib_size)) +
  geom_bar(stat = "identity") +
  scale_y_log10() +
  theme_bw()
```


### Read processing: evolution of library sizes

```{r}
run$lib_sizes %>%
  group_by(step) %>%
  summarize(n_reads = sum(lib_size))
```


## Separate samplenames into columns

```{r}
run$samples <- run$samples %>%
  separate(description2, into = c('participant', 'niche', 'swab', 'method')) %>%
  arrange(niche)
#View(run$samples)
```

change content of column names
```{r}
run_rename <- run %>%
  mutate(method = recode(method,
        "A" = "QIAamp Powerfecal DNA kit",
        "B" = "PMA + QIAamp Powerfecal DNA kit",
        "C" = "QIAamp DNA Microbiome kit",
        "D" = "HostZERO Microbial DNA kit")) %>%
  mutate(swab = recode(swab,
        "1" = "eNAT",
        "2" = "MSwab")) %>%
  mutate(niche = recode(niche,
        "V" = "Vagina",
        "SK" = "Skin",
        "SL" = "Saliva")) 
#View(run_rename$samples)

run <- run_rename
  #View(run$samples)
```


## Explore contaminats indicated by blanks

```{r,fig.height=3}
run_blanks <- 
  filter_samples(run, niche %in% c("PCR", "KIT"))
get_bar_plot(run_blanks, x = niche)
#View(run_blanks$samples)
```

*Supplementary Figure 8*

```{r}
ggplot(run$samples, aes(x = sample_id, y = concentration)) +
  geom_jitter(size=2, aes(colour=NC_kit))+
  geom_boxplot (alpha = 0.1) +
  scale_y_log10() +
  theme_bw()+
  theme(legend.position = "right", legend.text = element_text(size = 13)) +
  theme(axis.title=element_text(size=13), axis.text=element_text(size=13)) +
  labs(y= "Concentration (library size/pooled volume)") +
  labs(x= "Sample_id") 
```

Explore mock communities
```{r,fig.height=3}
run_MC <- run %>%
  filter_samples(pilot_study == "MC")
run_MC %>%
  aggregate_taxa(rank = "genus") %>%
  bar_plot()
#View(run_MC$taxa)
```


## Create a separate dataset per result sections

*Enrichment methods differ in discriminatory power of vagina, skin and saliva samples*
Dataset with 8 participants, 2 swabs, 4 techniques, 3 niches (192 samples)
```{r}
run_PS_1.0 <- run %>%
  filter_samples(pilot_study == "PS1.0") 
#View(run_PS_1.0$samples)
```

*Lysis buffer conserved microbial profiles over longer storage*
Dataset with 5 participants, 2 swabs, 1 technique, 1 niche, 4 timepoints (40 samples)
```{r}
run_PS_2.0 <- run %>%
  filter_samples(pilot_study == "PS2.0")
#View(run_PS_2.0$samples)
```

*Host DNA was effectively depleted for shotgun metagenomic sequencing*
*Taxonomic bias of enrichment methods towards Gram-negative taxa*
Dataset with 15 participants, 1 swab, 2 techniques, 3 niches (90 samples)
```{r}
run_PS_3.0 <- run %>%
  filter_samples(pilot_study == "PS3.0")
#View(run_PS_3.0$samples)
```


## Lysis buffer conserved microbial profiles over longer storage

*Figure 3*

```{r,fig.height=5, fig.width=12}
run_PS_2.0%>%
  mutate_samples(unique_id=paste(participant, niche,swab,method,sep="_"))%>%
  add_taxon_name()%>%
get_bar_plot(x = method) +
  facet_wrap(~participant,scales="free",ncol=1) +
  facet_grid(rows = NULL, cols = vars(participant), scales="free", space = "free", facets = vars(swab)) +
  scale_color_brewer(palette = "Paired")+
  theme_bw()+
  theme(legend.position = "right", legend.text = element_text(size = 14), legend.title = element_text(size = 14),axis.title.x = element_blank(), axis.title=element_text(size=14), axis.text=element_text(size=14),strip.text=element_text(size=14))
```

*Supplementary Figure 4*

```{r}
run_PS_2.0%>%
  betas()%>%
  filter(participant_1 == participant_2)%>%
  filter(swab_1==swab_2)%>%
  filter(method_1 != method_2)%>%
  filter(method_1 =="T1")%>%
  ggplot(aes(x=participant_1,y=beta,fill=swab_2))+
  geom_boxplot()+
  theme_bw()+
  theme(legend.position = "right", legend.text = element_text(size = 18)) +
  theme(axis.title=element_text(size=18), axis.text=element_text(size=18)) +
  labs(y= "inverse_simpson") +
  theme(axis.title.x = element_blank()) +
  scale_fill_brewer(palette = "Paired") 
```

*Supplementary Figure 5*

```{r}
run_PS_2.0 %>%
  add_alphas() %>%
  samples() %>%
  ggplot(aes(x =  swab, y = inverse_simpson, fill = method))+
  geom_boxplot() +
  #geom_jitter(width = 0.3, height = 0) +
  theme_bw()+
  theme(legend.position = "right", legend.text = element_text(size = 18)) +
  theme(axis.title=element_text(size=18), axis.text=element_text(size=18)) +
  labs(y= "inverse_simpson") +
  theme(axis.title.x = element_blank()) +
  scale_fill_brewer(palette = "Paired") 
```


## Impact of enrichment methods on clustering of 96 vagina, skin and saliva samples based on 16S rRNA amplicon sequencing data

*Figure 4A*

Take subset of 30 random samples for an explorative overview
```{r}
run_sub <-   
  run_PS_1.0 %>%
  filter_samples(swab == "eNAT") %>%
  filter_samples(1:n() <= 30)
#View(run_sub$samples)
```

Create the dendrogram object (as a tree)
```{r}
dendro <- 
  run_PS1_enat %>%
  get_rel_abundance_matrix() %>%
  vegdist() %>%
  hclust(method = "average") %>% 
  as.phylo()
```

Visualize the dendrogram
```{r}
dendro %>%
  ggtree(layout = "circular") %>%
  (run_subset$samples %>% 
     relocate(sample_id)) +
  geom_tiplab(aes(label = participant), size = 4, offset = 0.01) +
  geom_tippoint(aes(col = method), size = 3, shape = 1) +
  scale_color_brewer(palette = "Paired") +
  theme(legend.position = "bottom") +
  xlim(c(0,0.6))
```

Save the figure
```{r}
ggsave("results/dendrogram.png", units = "cm", width = 10, height = 14)
```

Save the figure to upload to Evolview
```{r}
write.tree(dendro,file = "..results/dendro_evol") #export dendrogram

write.csv(run_PS1_enat$samples, "../doc/run_dendro.csv") #export supporting file 
```

*Figure 4B*

```{r,fig.height=6, fig.width=8}
my_comparisons_1 <- list(c("PMA + QIAamp Powerfecal DNA kit","QIAamp DNA Microbiome kit"), c("PMA + QIAamp Powerfecal DNA kit","HostZERO Microbial DNA kit"), c("QIAamp DNA Microbiome kit","HostZERO Microbial DNA kit"))

run_PS1_enat%>%
  betas()%>%
  filter(participant_1 == participant_2)%>%
  filter(niche_1 == niche_2)%>%
  filter(method_1 != method_2)%>%
  filter(method_1 =="QIAamp Powerfecal DNA kit")%>%
  mutate(niche_1=factor(niche_1,levels=c("Vagina", "Skin", "Saliva")))%>%
  mutate(method_2 = factor(method_2, levels = c("PMA + QIAamp Powerfecal DNA kit", "QIAamp DNA Microbiome kit", "HostZERO Microbial DNA kit"))) %>%
  ggplot(aes(x=method_2,y=beta,fill=method_2))+
  facet_wrap(~niche_1,scales="fixed") +
  geom_boxplot(outlier.shape = NA)+
  stat_compare_means(comparisons = my_comparisons_1, test="wilcox.test")+
  theme_bw()+
  theme(legend.position = "bottom", legend.text = element_text(size = 10)) +
  theme(axis.title=element_text(size=18), axis.text=element_text(size=10)) +
  labs(y= "Bray-Curtis dissimilarity") +
  theme(axis.text.x = element_blank(),axis.title.x = element_blank(),strip.text=element_text(size=18)) +
  scale_fill_brewer(palette = "Paired") 
```


--------------------------------------------------------------------------------------------------------------------------------------------------
PART 2
--------------------------------------------------------------------------------------------------------------------------------------------------

## Data preparation

We start by loading the data. All data is bundled in a "tidyamplicons" object: it contains the read counts of all taxa in all samples, the metadata of the samples (sampling location, swab, method,...) and metadata of the taxa (taxonomic classification and interpretable taxon names). 
```{r setup, include=FALSE}
load("../data/isala_shotgun.rda")
```

```{r}
run_shotgun <- run
#View(run_shotgun$samples)

run_shotgun$samples<-run_shotgun$samples%>%
  separate(sample_2,into=c("participant","niche","method"),remove=F)

run_bacterial<- run_shotgun %>%
  filter_samples(method != "NA")%>%
  filter_taxa(domain =="d__Bacteria")

run_shotgun <- run_shotgun %>%
  filter_samples(method != "NA")%>%
  filter_samples(niche != "KIT")
 #should be 86 samples (minus 4 skin samples)
```


### Quality control

```{r read depth, echo = F}
lib_size_without_human <- run_shotgun %>%
  filter_taxa(domain == "d__Bacteria")%>%
  abundances()%>%
  group_by(sample_id)%>%
  summarise(lib_size_without_human_reads = sum(abundance))

#get a table of the libsizes
run_shotgun%>%
  add_lib_size()%>%
  samples()%>%
  left_join(lib_size_without_human)%>%
  mutate(lib_size_with_human_reads=lib_size)%>%
  select(-lib_size,-run)%>%
  write_csv("tabel_libsizes.csv")

#change content of column names
run_rename <- run_shotgun %>%
  mutate_samples(method=ifelse(method=="A","QIAamp Powerfecal DNA kit", method))%>%
  mutate_samples(method=ifelse(method=="D","HostZERO Microbial DNA kit", method))%>%
  mutate_samples(niche=ifelse(niche=="V","Vagina", niche))%>%
  mutate_samples(niche=ifelse(niche=="SK","Skin", niche))%>%
  mutate_samples(niche=ifelse(niche=="SL","Saliva", niche))
#View(run_rename$samples)

run_shotgun <- run_rename
#View(run_shotgun$samples)
```

Next we calculate the library concentration by making a new column: lib_size/pooled volume = concentration. But don't forget to first add lib_size
```{r}
run_shotgun <- add_lib_size(run_shotgun)
run_shotgun <- mutate_samples (run_shotgun, concentration = lib_size / pooled_volume)
#View(run_shotgun$samples)
```


## Host DNA was effectively depleted for shotgun metagenomic sequencing

*Figure 5*

```{r}
run_shotgun_1 <- run_shotgun %>%
  filter_taxa(domain == "d__Bacteria")
#View(run_shotgun_1$taxa)

run_shotgun_1 <- run_shotgun_1 %>%
  add_lib_size()%>% 
  mutate_samples (bacterial_proportion = lib_size / lib_size_with_human_reads)
```

```{r,fig.height=5, fig.width=14}
my_comparisons_1 <- list(c("QIAamp Powerfecal DNA kit", "HostZERO Microbial DNA kit"))

run_shotgun_1%>%
  samples() %>%
  mutate(method=factor(method,levels=c("QIAamp Powerfecal DNA kit", "HostZERO Microbial DNA kit")))%>%
  mutate(niche=factor(niche,levels=c("Vagina", "Skin", "Saliva")))%>%
  ggplot(aes(x=method,y=bacterial_proportion,fill=method))+
  geom_jitter(alpha=1,width=0.1,height = 0,aes(col=method), cex = 2)+
  facet_wrap(~niche)+
  stat_summary(fun=mean, geom="errorbar", aes(ymax = ..y.., ymin = ..y..), width = .76, linetype = "dashed")+
  stat_compare_means(comparisons = my_comparisons_1, test="wilcox.test")+
  theme_bw()+
  theme(legend.position = "right", legend.text = element_text(size = 18), legend.title = element_text(size = 18), axis.title=element_text(size=18), axis.text=element_text(size=10),axis.text.x = element_blank(),axis.title.x = element_blank(),strip.text=element_text(size=18))
```

*Supplementary Figure 5*

```{r}
run_shotgun_test <- run_shotgun_1 %>%
  add_lib_size()%>% 
  mutate_samples (bacterial_read_concentration = lib_size / pooled_volume)%>% 
  samples()
#View(run_shotgun_test)
```

```{r,fig.height=6, fig.width=9}
run_shotgun_test%>%
  #samples() %>%
  mutate(method=factor(method,levels=c("QIAamp Powerfecal DNA kit", "HostZERO Microbial DNA kit")))%>%
  mutate(niche=factor(niche,levels=c("Vagina", "Skin", "Saliva")))%>%
  ggplot(aes(x=method,y=bacterial_read_concentration,fill=method))+
  #geom_boxplot(outlier.shape = NA)+
  geom_jitter(alpha=1,width=0.1,height = 0,aes(col=method))+
  facet_wrap(~niche)+
  #scale_y_continuous()+
  scale_x_discrete(guide = guide_axis(n.dodge=3))+
  labs(title="Dodge Overlapping X-axis Label Text\nwith ggplot2 3.3.0") +
  scale_y_log10()+
  theme_bw()+
  theme(legend.position = "bottom")
```


## Taxonomic bias of enrichment methods towards Gram-negative taxa

*Figure 6*

Vagina
```{r}
Vagina_genus <- run_Vagina %>% aggregate_taxa(rank = "genus") 

#View(Vagina_genus$samples)
#View(Vagina_genus$taxa)

Vagina_genus %>%
  mutate_taxa(genus=ifelse(genus == "s__Peptostreptococcaceae bacterium oral taxon 929", "g_taxon 929", genus))%>%
  mutate_taxa(gram = if_else(phylum %in% c("p__Firmicutes", "p__Actinobacteria"),"(+)","(-)")) %>% 
  add_taxon_name() %>%
  mutate_taxa(taxon_name = str_c(taxon_name, gram, sep = " ")) %>% 
  add_codifab(method, conditions = c("HostZERO_Microbial_DNA_kit","QIAamp_Powerfecal_DNA_kit"), max_taxa = 30) %>%
  codifab_plot(HostZERO_Microbial_DNA_kit_vs_QIAamp_Powerfecal_DNA_kit) 

save(Vagina_genus, file = "vagina_genus.rda")
```

Skin
```{r}
Skin_genus <- run_Skin %>% aggregate_taxa(rank = "genus") 

#View(Skin_genus$samples)
#View(Skin_genus$taxa)

Skin_genus %>%
  mutate_taxa(gram = if_else(phylum %in% c("p__Firmicutes", "p__Actinobacteria"),"(+)","(-)")) %>% 
  add_taxon_name() %>%
  mutate_taxa(taxon_name = str_c(taxon_name, gram, sep = " ")) %>% 
  add_codifab(method, conditions = c("HostZERO_Microbial_DNA_kit","QIAamp_Powerfecal_DNA_kit"), max_taxa = 30) %>%
  codifab_plot(HostZERO_Microbial_DNA_kit_vs_QIAamp_Powerfecal_DNA_kit) 
```

Saliva
```{r}
Saliva_genus <- run_Saliva %>% aggregate_taxa(rank = "genus") 

#View(Saliva_genus$samples)
#View(Saliva_genus$taxa)

Saliva_genus %>%
  mutate_taxa(gram = if_else(phylum %in% c("p__Firmicutes", "p__Actinobacteria"),"(+)","(-)")) %>% 
  add_taxon_name() %>%
  mutate_taxa(taxon_name = str_c(taxon_name, gram, sep = " ")) %>% #plak 2 variabelen aan elkaar
  add_codifab(method, conditions = c("HostZERO_Microbial_DNA_kit","QIAamp_Powerfecal_DNA_kit"), max_taxa = 30) %>%
  codifab_plot(HostZERO_Microbial_DNA_kit_vs_QIAamp_Powerfecal_DNA_kit) 
```

*Supplementary Figure 7*

```{r,fig.height=6, fig.width=8}
run_shotgun_1%>%
  betas()%>%
  filter(participant_1 == participant_2)%>%
  filter(niche_1 == niche_2)%>%
  filter(method_1 != method_2)%>%
  filter(method_1 =="QIAamp Powerfecal DNA kit")%>%
  mutate(niche_1=factor(niche_1,levels=c("Vagina", "Skin", "Saliva")))%>%
  mutate(method_2 = factor(method_2, levels = c("HostZERO Microbial DNA kit"))) %>%
  ggplot(aes(x=niche_1,y=beta,fill=method_2))+
  geom_boxplot(outlier.shape = NA)+
  theme_bw()+
  theme(legend.position = "bottom", legend.text = element_text(size = 18)) +
  theme(axis.title=element_text(size=18), axis.text=element_text(size=18)) +
  labs(y= "Bray-Curtis dissimilarity") +
  theme(axis.title.x = element_blank()) +
  scale_fill_brewer(palette = "Paired") 
```

__________________________________________________________________________________________________________________________________________________

The end
__________________________________________________________________________________________________________________________________________________

